### PR TITLE
Target deduplication and deterministic scraping times

### PR DESCRIPTION
**Pull request by fabxc**
Cleaning up some debt, removing a lot of fragility.

Targets are now immutable. This means no more fragile updating whenever we get a new target group from SD. No more convoluted logic of starting and stopping scrapers.
All targets are thrown into a `ScrapePool` that does the deduping and maintains the state we actually want to keep, which is very little. On changes it drops all targets, stops all scrapers and restarts with the new (deduped) list of targets.

@juliusv

### CODE REVIEW

File: retrieval/scrape.go [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+		sp.cancel()
+	}
+}
+
+// Sync the running scrapers with the provided list of targets.
+func (sp *ScrapePool) Sync(targets []*Target) {
+	sp.mtx.Lock()
+	defer sp.mtx.Unlock()
+
+	// Shutdown all running loops after their current scrape
```
Comment (brian-brazil) [inline retrieval/scrape.go]: Pausing all scraping every time there's an update seems a bit extreme, it's conceivable that there'll be setups that'll have multiple target group updates per second. I think we need to handle this a bit more gracefully.

Comment (fabxc) [inline retrieval/scrape.go]: That's exactly the point of all of this. We want to keep this simple.

What happens here is that we stop all scrape loops. They do not terminate until they have successfully finished any ongoing scrapes.
Before those terminate we already start all the new loops. With predictable slotting the first scrape of the new loops will be exactly the right interval after the last scrape of the old loops.

There is a tiny gap between *initiating* all the stops and immediately starting all the new loops. Should this really be an issue, i.e. targets frequently missing scrapes, we can place a timer and let the new loops start scrapes the fell into the gap anyway.

Simplifying the design here for a few more irrelevant allocations and cycles is well worth it.

Comment (fabxc) [inline retrieval/scrape.go]: The calls to `Sync()` only happen in one second intervals by the target manager if there were any updates. So yes, that's "worse" in theory than before. Bugs are worse.

Comment (brian-brazil) [inline retrieval/scrape.go]: Ah, I misunderstood the logic.

> With predictable slotting the first scrape of the new loops will be exactly the right interval after the last scrape of the old loops.

There's still a random offset added to each scrape loop, at the least we'd need to change this to something based on the hash of the target to make this work.

Comment (fabxc) [inline retrieval/scrape.go]: That's why there's a [WIP] tag and a 'missing' section :)

File: retrieval/scrape.go [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/log"
+	"golang.org/x/net/context"
+
+	"github.com/prometheus/prometheus/storage"
+)
+
+// Scraper retrieves samples.
+type Scraper interface {
```
Comment (juliusv) [inline retrieval/scrape.go]: Unexport these types if they're not used outside of the package.

File: retrieval/scrape.go [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+	// Sends samples to ch until an error occurrs, the context is canceled
+	// or no further samples can be retrieved. The channel is closed before returning.
+	Scrape(ctx context.Context, ch chan<- *model.Sample) error
+}
+
+// Scrape pools runs scrapers for a set of targets where a target, identified by
+// a label set is scraped exactly once.
```
Comment (juliusv) [inline retrieval/scrape.go]: s/label set/label set,/

File: retrieval/scrape.go [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+
+	for _, l := range sp.loops {
+		go func(l *scrapeLoop) {
+			l.stop()
+			wg.Done()
+		}(l)
+	}
+	defer wg.Wait()
+
+	// Old loops are shutting down, create new from the targets.
```
Comment (juliusv) [inline retrieval/scrape.go]: s/new/new ones/

File: retrieval/scrape.go [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+	select {
+	case <-time.After(time.Duration(float64(interval) * rand.Float64())):
+		// Continue after random scraping offset.
+	case <-ctx.Done():
+		return
+	case <-sl.stopc:
+		return
+	}
+
+	timer := time.NewTicker(interval)
```
Comment (juliusv) [inline retrieval/scrape.go]: I'd call it `ticker` to not cause confusion because the stdlib also differentiates between Timers and Tickers.

Comment (fabxc) [inline retrieval/scrape.go]: That was a timer for a short time, some leftovers remained.

**Comment by juliusv** [comment]
Got interrupted a billion times trying to look at this today, but the approach looks good in general. Will give it a deeper look once it's no longer WIP...

**Comment by fabxc** [comment]
Tests added, deterministic scraping added.

@juliusv @brian-brazil

File: retrieval/scrape.go [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+		return
+	}
+
+	ticker := time.NewTicker(interval)
+	defer ticker.Stop()
+
+	for {
+		start := time.Now()
+
+		err := sl.scrape(ctx, timeout)
```
Comment (brian-brazil) [inline retrieval/scrape.go]: If timeout > interval, the new logic could result in two scrapes in progress at once which won't end well. We should probably require that timeout <= interval.

Comment (fabxc) [inline retrieval/scrape.go]: Ah yes, discussed this with @juliusv already. 
That requirement should be enforced via config validation – forgot to add that.

File: retrieval/target.go [hunk 7]
```diff
@@ -88,11 +89,8 @@ func (t TargetHealth) value() model.SampleValue {
 }
 
 const (
-	// HealthUnknown is the state of a Target before it is first scraped.
```
Comment (grobie) [inline retrieval/target.go]: Isn't golint complaining if you remove all doc strings? I think the minimum required is a signle line above the `const` keyword.

File: retrieval/targetmanager.go [hunk 8]
```diff
@@ -187,6 +202,14 @@ func (tm *TargetManager) handleUpdates(ch <-chan targetGroupUpdate, done <-chan
 			if err := tm.updateTargetGroup(update.tg, update.scfg); err != nil {
 				log.Errorf("Error updating targets: %s", err)
 			}
+			dirty = true
+
+		case <-syncTicker.C:
+			if dirty {
+				tm.sync()
```
Comment (brian-brazil) [inline retrieval/targetmanager.go]: We should probably add a counter here, and on each update

**Comment by brian-brazil** [comment]
:+1:

File: retrieval/target.go (hunk 7 above)
Comment (fabxc) [inline retrieval/target.go]: Probably – single line above `const` sounds good though. Just not that excited by doc strings consisting of spaced combinations of identifier names.

Comment (grobie) [inline retrieval/target.go]: "Close closes the connection"

Comment (fabxc) [inline retrieval/target.go]: Right after `// XYZ implements the XYZer interface` – why do we even have implicit interfaces?

File: retrieval/scrape.go [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+	}
+	durationSample := &model.Sample{
+		Metric: model.Metric{
+			model.MetricNameLabel: scrapeDurationMetricName,
+		},
+		Timestamp: ts,
+		Value:     model.SampleValue(float64(duration) / float64(time.Second)),
+	}
+
+	sl.appender.Append(healthSample)
```
Comment (brian-brazil) [inline retrieval/scrape.go]: Metric relabelling shouldn't apply to the non-scraped samples. I thought I'd a unittest for that.

Comment (fabxc) [inline retrieval/scrape.go]: Done.

**Comment by brian-brazil** [comment]
:+1:

File: retrieval/target.go [hunk 10]
```diff
@@ -451,33 +338,23 @@ func (t *Target) scrape(appender storage.SampleAppender) (err error) {
 	sdec := expfmt.SampleDecoder{
 		Dec: dec,
 		Opts: &expfmt.DecodeOptions{
-			Timestamp: model.TimeFromUnixNano(start.UnixNano()),
+			Timestamp: model.TimeFromUnixNano(time.Now().UnixNano()),
```
Comment (juliusv) [inline retrieval/target.go]: Previously the scrape timestamp was always the very start of the scrape, which we assumed to be the most accurate guess for most use cases (the metrics are all snapshotted at the beginning of the request, then transferred). Any particular reason why this changes the timestamp to the beginning of the body reception?

Comment (fabxc) [inline retrieval/target.go]: Naah.

File: retrieval/scrape.go [hunk 11]
```diff
@@ -0,0 +1,294 @@
...
+			// TODO(fabxc): Change the SampleAppender interface to return an error
+			// so we can proceed based on the status and don't leak goroutines trying
+			// to append a single sample after dropping all the other ones.
+			sl.appender.Append(smpl)
+		}
+	}()
+
+	ctx, _ = context.WithTimeout(ctx, timeout)
+
+	return sl.scraper.Scrape(ctx, ch)
```
Comment (juliusv) [inline retrieval/scrape.go]: We might want to avoid passing every single sample through a channel here. We did that once in the past (where admittedly channel sends were even much slower), but then noticed that it slowed everything to a crawl, then switched to sending batches of samples through channels. The current HEAD does just synchronous calls, I think.

Just benchmarking this it's 96ns per sample (with channel cap of 256) for the channel send vs. 0.31ns when just passing it to a function. For 800k samples/s, that'd be 77ms per second used on the channel send here. Can we just pass in the appender to `Scrape()` and then do something like this there:

```
		for _, smpl := range samples {
			select {
			case <-ctx.Done():
				return ctx.Err()
			default:
				app.Append(smpl)
			}
 		}
```

File: retrieval/scrape_test.go [hunk 12]
```diff
@@ -0,0 +1,542 @@
...
+	if !reflect.DeepEqual(app, sp.appender) {
+		t.Fatalf("Wrong sample appender")
+	}
+
+	if sp.ctx == nil || sp.cancel == nil {
+		t.Fatalf("Context not initialized")
+	}
+
+	if sp.newLoop == nil {
+		t.Fatalf("newLoop function not initializated")
```
Comment (juliusv) [inline retrieval/scrape_test.go]: s/initializated/initialized/

File: retrieval/scrape.go (hunk 11 above)
Comment (juliusv) [inline retrieval/scrape.go]: Btw., with Go 1.4.2, it was just 71ns, but Go 1.5.1 raised it to 96ns per sample.

File: retrieval/scrape_test.go [hunk 13]
```diff
@@ -0,0 +1,542 @@
...
+		sl.reportAppender = appender
+
+		sl.report(test.start.Time(), test.duration, test.err)
+
+		if sl.state.LastError() != test.err {
+			t.Fatalf("Expected error %q but got %q", test.err, sl.state.LastError())
+		}
+
+		if sl.state.LastScrape() != test.start.Time() {
+			t.Fatalf("Expeceted test start at %v but got %v", test.start.Time(), sl.state.LastScrape())
```
Comment (juliusv) [inline retrieval/scrape_test.go]: s/Expeceted/Expected/

File: retrieval/scrape_test.go [hunk 14]
```diff
@@ -0,0 +1,542 @@
...
+		select {
+		case <-appender.ch:
+		case <-time.After(timeout):
+			t.Fatalf("Timeout on expecting sample %d", i)
+		}
+	}
+
+	select {
+	case <-appender.ch:
+		t.Fatalf("Received more sample than sent by the scraper")
```
Comment (juliusv) [inline retrieval/scrape_test.go]: s/sample/samples/

**Comment by juliusv** [comment]
:+1: otherwise

File: retrieval/scrape.go (hunk 11 above)
Comment (fabxc) [inline retrieval/scrape.go]: Valid point regarding performance, don't think this is a good solution though.
The point of this PR is fixing all the logic leaks, i.e. scrape loops worrying about wrapping samples appenders with relabeling, target managers worrying about scraping, etc.

Similarly a `Scrape()` function shouldn't know such a thing as a `SampleAppender`.

Making the channel `chan<- []*model.Sample` should allow us to maintain separation here without significant performance impact.

Comment (juliusv) [inline retrieval/scrape.go]: That sounds good to me too. That's what we once used to have and it worked well.

Comment (fabxc) [inline retrieval/scrape.go]: The point with that is, again, that we cannot reuse the slices – but hasn't been a problem so far either.

**Comment by fabxc** [comment]
All done.

**Comment by beorn7** [comment]
This PR is very invasive. It deals with the carefully crafted and optimized ingestion pipeline. I don't have gained enough insight yet to estimate the impact. I recommend to benchmark it excessively, i.e. set up a server with a very high ingestion rate (350k samples / sec) and make sure it can sustain that rate for hours.

**Comment by fabxc** [comment]
That is currently happening.

**Comment by fabxc** [comment]
Superseded by #1404
